File: backend/app/schemas/agents.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class AgentIdentity(_StrictModel):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)


class AgentInstructions(_StrictModel):
    role: str = Field(default="", max_length=200)
    objective: str = Field(default="", max_length=2000)
    system_prompt: str = Field(default="", max_length=8000)
    greeting: str = Field(default="", max_length=1000)
    closing: str = Field(default="", max_length=1000)


class AgentBehavior(_StrictModel):
    response_style: Literal["precise", "balanced", "creative"] = "balanced"
    interruptions: Literal["conservative", "balanced", "responsive"] = "balanced"
    turn_detection: Literal["automatic", "conservative", "balanced", "responsive"] = "automatic"
    confirmation_strategy: Literal["important_data", "always", "never"] = "important_data"
    agent_first: bool = True


class ProviderAgentReference(_StrictModel):
    agent_id: str = Field(min_length=1, max_length=120)
    observed_published_revision_id: str | None = Field(default=None, max_length=120)


_FORBIDDEN_VOICE_SETTINGS_KEY_PARTS = (
    "api_key", "apikey", "secret", "token", "password", "authorization", "header",
)


def _reject_secret_keys(value: dict[str, Any]) -> dict[str, Any]:
    if any(part in str(key).lower() for key in value for part in _FORBIDDEN_VOICE_SETTINGS_KEY_PARTS):
        raise ValueError("Settings contain a forbidden secret field")
    return value


class AgentVoiceConfig(_StrictModel):
    """A voice selection for a realtime agent.

    `mode="provider"` is a voice known/managed by the realtime provider's own
    catalog (e.g. an Ultravox voice) -- it does not imply the provider
    synthesizes the audio itself, only that ServiGlobal doesn't need to know
    which TTS backs it. `mode="provider_external"` asks the realtime provider
    to delegate synthesis to a named external TTS provider (e.g. Ultravox's
    externalVoice -> ElevenLabs) using an ID from that provider's own
    namespace, never the realtime provider's voice IDs.

    This schema validates shape only: `mode`/`provider`/`voice_id` types and
    that `settings` is a plain mapping with no secret-like keys. It
    deliberately does NOT know which settings keys or ranges a given
    (mode, provider) actually allows -- that is provider-specific business
    logic and lives in VoiceSelectionService, which has the sibling realtime
    provider/model context this schema doesn't, and is the single place
    that logic should be read or changed.
    """

    mode: Literal["provider", "provider_external"] = "provider"
    provider: str = Field(max_length=40)
    voice_id: str = Field(min_length=1, max_length=160)
    settings: dict[str, Any] = Field(default_factory=dict)

    @field_validator("settings")
    @classmethod
    def reject_secret_settings(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _reject_secret_keys(value)


class AgentToolBinding(_StrictModel):
    """One tool bound to a serviglobal_managed agent version. `key` is
    validated against the platform Tool Registry (app.domain.tool_registry)
    by AgentService, not here -- this schema validates shape only: key
    length, no duplicate enforcement (that needs the whole list), and no
    secrets in `config`. `config` is binding-time configuration (e.g. which
    WhatsApp template to use), never the per-call arguments a tool
    invocation carries -- those come from the LLM at call time and are
    validated against ToolDefinition.input_schema separately.
    """

    key: str = Field(min_length=1, max_length=80)
    enabled: bool = True
    config: dict[str, Any] = Field(default_factory=dict)

    @field_validator("config")
    @classmethod
    def reject_secret_config(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _reject_secret_keys(value)


class ProviderManagedOverrides(_StrictModel):
    join_timeout: str | None = Field(default=None, max_length=24)
    max_duration: str | None = Field(default=None, max_length=24)
    recording_enabled: bool | None = None
    initial_output_medium: Literal["MESSAGE_MEDIUM_VOICE", "MESSAGE_MEDIUM_TEXT"] | None = None


class _RuntimeSelection(_StrictModel):
    management_mode: Literal["serviglobal_managed", "provider_managed"] = "serviglobal_managed"
    provider_agent: ProviderAgentReference | None = None
    voice: AgentVoiceConfig | None = None
    provider_overrides: ProviderManagedOverrides | None = None
    # Provider/model runtime parameters (e.g. temperature) for a
    # serviglobal_managed agent. Shape-only here (no secrets); which keys
    # are actually allowed for a given (provider, model) is registry-driven
    # business logic, validated by AgentService against
    # voice_registry.validate_model_settings(), not by this schema.
    settings: dict[str, Any] = Field(default_factory=dict)
    # Tools bound to this agent version. Which keys are known/executable
    # comes from app.domain.tool_registry, validated by AgentService --
    # this schema only validates shape (see AgentToolBinding).
    tools: list[AgentToolBinding] = Field(default_factory=list)

    @field_validator("settings")
    @classmethod
    def reject_secret_runtime_settings(cls, value: dict[str, Any]) -> dict[str, Any]:
        return _reject_secret_keys(value)

    @model_validator(mode="after")
    def validate_management_mode(self):
        if self.management_mode == "provider_managed" and self.provider_agent is None:
            raise ValueError("provider_agent is required for provider_managed agents")
        if self.management_mode == "serviglobal_managed" and self.provider_agent is not None:
            raise ValueError("provider_agent is only valid for provider_managed agents")
        return self


class AgentCreateRequest(_RuntimeSelection):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    language: str = Field(default="es", min_length=2, max_length=16)
    timezone: str = Field(default="America/Bogota", max_length=80)
    instructions: AgentInstructions = AgentInstructions()
    behavior: AgentBehavior = AgentBehavior()
    voice_agent_config_id: str | None = Field(default=None, max_length=36)
    pipeline_type: Literal["realtime"] = "realtime"
    provider: str = Field(default="ultravox", max_length=40)
    model: str = Field(default="ultravox", max_length=80)


class AgentUpdateRequest(_StrictModel):
    """Legacy identity-only update, kept for API compatibility.

    The Agent Builder UI no longer uses this: it PATCHes name/description
    together with the rest of the draft via AgentDraftUpdateRequest so both
    land in one transaction. This endpoint still works standalone for any
    other caller, but on a mutable agent with an open draft it will leave
    that draft's identity_json snapshot stale until the draft is next saved.
    """

    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)


class AgentDraftUpdateRequest(_RuntimeSelection):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    language: str = Field(default="es", min_length=2, max_length=16)
    timezone: str = Field(default="America/Bogota", max_length=80)
    instructions: AgentInstructions = AgentInstructions()
    behavior: AgentBehavior = AgentBehavior()
    voice_agent_config_id: str | None = Field(default=None, max_length=36)
    pipeline_type: Literal["realtime"] = "realtime"
    provider: str = Field(default="ultravox", max_length=40)
    model: str = Field(default="ultravox", max_length=80)


class AgentPublishRequest(_StrictModel):
    """Optional concurrency guard: when set, publish fails with a conflict
    if the agent's current draft_version_id no longer matches (someone else
    saved or published a different draft in the meantime)."""

    expected_draft_version_id: str | None = None


class AgentToolCatalogEntryResponse(_StrictModel):
    """One Tool Registry entry, annotated for the current tenant.
    `available` is only ever true for `status="available"` tools whose
    `required_integration` is actually configured for this tenant --
    `planned` tools always report `available=False` so the Agent Builder
    UI never lets them be enabled."""

    key: str
    name: str
    description: str
    status: Literal["available", "planned"]
    required_integration: Literal["booking", "whatsapp", "crm", "chatwoot"] | None
    available: bool
    input_schema: dict[str, Any]
    source: Literal["platform", "custom"] = "platform"


class AgentResponse(_StrictModel):
    id: str
    name: str
    description: str | None
    status: Literal["draft", "active", "archived"]
    published_version_id: str | None
    draft_version_id: str | None
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AgentVersionResponse(_StrictModel):
    id: str
    agent_id: str
    version: int
    status: Literal["draft", "published", "superseded"]
    language: str
    timezone: str
    identity: AgentIdentity
    instructions: AgentInstructions
    behavior: AgentBehavior
    runtime_binding: dict
    voice_agent_config_id: str | None
    published_at: datetime | None
    created_at: datetime

File: frontend/types/agents.ts
export type AgentStatus = 'draft' | 'active' | 'archived';
export type AgentVersionStatus = 'draft' | 'published' | 'superseded';
export type AgentResponseStyle = 'precise' | 'balanced' | 'creative';
export type AgentInterruptions = 'conservative' | 'balanced' | 'responsive';
export type AgentTurnDetection = 'automatic' | 'conservative' | 'balanced' | 'responsive';
export type AgentConfirmationStrategy = 'important_data' | 'always' | 'never';

export type AgentIdentity = {
  name: string;
  description?: string | null;
};

export type AgentInstructions = {
  role: string;
  objective: string;
  system_prompt: string;
  greeting: string;
  closing: string;
};

export type AgentBehavior = {
  response_style: AgentResponseStyle;
  interruptions: AgentInterruptions;
  turn_detection: AgentTurnDetection;
  confirmation_strategy: AgentConfirmationStrategy;
  agent_first: boolean;
};

/** Provider/model runtime parameters (e.g. temperature). Which keys are
 * valid for a given (provider, model) comes from VoiceModelResponse.parameters
 * (see types/voice-registry.ts) -- this type is intentionally a generic bag,
 * never hardcoded to one provider. */
export type AgentModelSettings = Record<string, number | boolean | string>;

/** A tool bound to a serviglobal_managed agent version. `config` is
 * binding-time configuration (e.g. whatsapp.send_message's V2 template/
 * recipient/variables contract -- see WhatsAppToolBindingConfig), never
 * the per-call arguments the LLM supplies at invocation time. */
export type AgentToolBinding = {
  key: string;
  enabled: boolean;
  config?: Record<string, unknown>;
};

export type AgentToolContextRequirement = {
  path: string;
  required: boolean;
  description: string;
};

/** One catalog entry (platform Registry or tenant custom.* tool) annotated
 * for the current tenant -- see GET /api/v1/agents/tools/catalog.
 * `available` is only ever true for status="available" tools whose
 * required_integration (platform) or credential (custom) is configured;
 * "planned" tools always report available=false and must never be offered
 * as selectable in the UI. `source` distinguishes the two: `custom` tools
 * always report `required_integration: null` -- their "is this tenant
 * ready" concept is credential-based, checked via /voice-ai/tools instead
 * of an integration settings page. `input_schema` is the tool's static/base
 * LLM schema -- for a `configuration_required` tool, the real schema shown
 * to the model depends on `config` and is computed server-side (see
 * PlatformToolContractService.compile_llm_schema on the backend); this
 * type only carries what the catalog can say ahead of any binding.
 * `context_requirements` documents what the tool reads from SessionContextV1
 * server-side -- informational only, never rendered as an editable field. */
export type AgentToolCatalogEntry = {
  key: string;
  name: string;
  description: string;
  status: 'available' | 'planned';
  required_integration: 'booking' | 'whatsapp' | 'crm' | 'chatwoot' | null;
  available: boolean;
  input_schema: Record<string, unknown>;
  source: 'platform' | 'custom';
  context_requirements: AgentToolContextRequirement[];
  binding_config_schema: Record<string, unknown>;
  configuration_required: boolean;
};

/** whatsapp.send_message's contract_version=2 AgentToolBinding.config shape
 * -- kept here (not enforced client-side beyond the configurator's own
 * form) because the backend (PlatformToolContractService) is the actual
 * source of truth for validity; this type exists so the configurator UI
 * and buildToolsPayload can build/read it without `as any`. */
export type WhatsAppVariableSource = 'llm' | 'context' | 'fixed';

export type WhatsAppVariableMapping =
  | { source: 'llm'; type: string; description?: string }
  | { source: 'context'; path: string }
  | { source: 'fixed'; value: string };

export const WHATSAPP_ALLOWED_CONTEXT_PATHS = [
  'caller.phone',
  'contact.id',
  'contact.name',
  'contact.phone',
  'contact.email',
  'lead.id',
  'lead.status',
  'lead.stage',
  'campaign.name',
] as const;

export type WhatsAppRecipientStrategy = 'contact_then_caller' | 'contact' | 'caller';

export type WhatsAppToolBindingConfig = {
  contract_version: 2;
  template_key: string;
  recipient: { strategy: WhatsAppRecipientStrategy };
  variables: Record<string, WhatsAppVariableMapping>;
};

export type AgentCreateRequest = {
  name: string;
  description?: string | null;
  language: string;
  timezone: string;
  instructions: AgentInstructions;
  behavior: AgentBehavior;
  voice_agent_config_id?: string | null;
  pipeline_type: 'realtime';
  provider: string;
  model: string;
  management_mode?: 'serviglobal_managed' | 'provider_managed';
  provider_agent?: ProviderAgentReference | null;
  voice?: AgentVoiceConfig | null;
  settings?: AgentModelSettings | Record<string, never>;
  tools?: AgentToolBinding[];
};

export type ProviderAgentReference = {
  agent_id: string;
  observed_published_revision_id?: string | null;
};

/** ElevenLabs is the only provider_external target with a real adapter in
 * V1 (see backend voice_registry.py); the settings keys below are exactly
 * the allowlist VoiceSelectionService enforces server-side. `model` is a
 * free string by design -- no closed enum, see Phase D decisions. */
export type ExternalVoiceSettings = {
  model?: string;
  speed?: number;
  stability?: number;
  similarity_boost?: number;
  use_speaker_boost?: boolean;
};

export type AgentVoiceConfig = {
  mode: 'provider' | 'provider_external';
  provider: string;
  voice_id: string;
  settings?: ExternalVoiceSettings | Record<string, never>;
};

export type AgentUpdateRequest = {
  name: string;
  description?: string | null;
};

export type AgentDraftUpdateRequest = {
  name: string;
  description?: string | null;
  language: string;
  timezone: string;
  instructions: AgentInstructions;
  behavior: AgentBehavior;
  voice_agent_config_id?: string | null;
  pipeline_type: 'realtime';
  provider: string;
  model: string;
  management_mode?: 'serviglobal_managed' | 'provider_managed';
  provider_agent?: ProviderAgentReference | null;
  voice?: AgentVoiceConfig | null;
  settings?: AgentModelSettings | Record<string, never>;
  tools?: AgentToolBinding[];
};

export type AgentPublishRequest = {
  expected_draft_version_id?: string | null;
};

export type AgentResponse = {
  id: string;
  name: string;
  description: string | null;
  status: AgentStatus;
  published_version_id: string | null;
  draft_version_id: string | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
};

export type AgentRuntimeBinding = {
  pipeline_type: 'realtime';
  realtime: {
    provider: string;
    model: string;
    management_mode?: 'serviglobal_managed' | 'provider_managed';
    provider_agent?: ProviderAgentReference | null;
    voice?: AgentVoiceConfig | null;
    settings?: AgentModelSettings | Record<string, never>;
    provider_extensions?: {
      observed_revision_drift?: boolean;
      tools?: Array<{ name: string; classification: string }>;
      warnings?: string[];
    };
  };
  // Top-level, sibling of `realtime` -- never nested inside it, so it can
  // never collide with the unrelated realtime.provider_extensions.tools
  // (a read-only classification of an imported provider_managed agent's
  // own remote tools).
  tools?: AgentToolBinding[];
};

export type AgentVersionResponse = {
  id: string;
  agent_id: string;
  version: number;
  status: AgentVersionStatus;
  language: string;
  timezone: string;
  identity: AgentIdentity;
  instructions: AgentInstructions;
  behavior: AgentBehavior;
  runtime_binding: AgentRuntimeBinding;
  voice_agent_config_id: string | null;
  published_at: string | null;
  created_at: string;
};

export type AgentGateState = 'ok' | 'feature_disabled' | 'access_denied' | 'error';
